public/app/plugins/datasource/loki/modifyQuery.ts

```diff
@@ -165,45 +166,45 @@ export function addLabelToQuery(
   const everyStreamSelectorHasMatcher = streamSelectorPositions.every((streamSelectorPosition) =>
     hasStreamSelectorMatchers.some(
       (matcherPosition) =>
         matcherPosition.from >= streamSelectorPosition.from && matcherPosition.to <= streamSelectorPosition.to
     )
   );
-  const parserPositions = getParserPositions(query);
-  const labelFilterPositions = getLabelFilterPositions(query);
 
   const filter = toLabelFilter(key, value, operator);
   // If we have non-empty stream selector and parser/label filter, we want to add a new label filter after the last one.
   // If some of the stream selectors don't have matchers, we want to add new matcher to the all stream selectors.
-  if (forceAsLabelFilter) {
-    // `forceAsLabelFilter` is mostly used for structured metadata labels. Those are not
-    // very well distinguishable from real labels, but need to be added as label
-    // filters after the last stream selector, parser or label filter. This is
-    // just a quickfix for now and still has edge-cases where it can fail.
-    // TODO: improve this once we have a better API in Loki to distinguish
-    // between the origins of labels.
+  if (labelType === LabelType.Parsed || labelType === LabelType.StructuredMetadata) {
     const positionToAdd = findLastPosition([...streamSelectorPositions, ...labelFilterPositions, ...parserPositions]);
     return addFilterAsLabelFilter(query, [positionToAdd], filter);
-  } else if (everyStreamSelectorHasMatcher && (labelFilterPositions.length || parserPositions.length)) {
-    // in case we are not adding the label to stream selectors we need to find the last position to add in each expression
-    const subExpressions = findLeaves(getNodePositionsFromQuery(query, [Expr]));
-    const parserFilterPositions = [...parserPositions, ...labelFilterPositions];
-
-    // find last position for each subexpression
-    const lastPositionsPerExpression = subExpressions.map((subExpression) => {
-      return findLastPosition(
-        parserFilterPositions.filter((p) => {
-          return subExpression.contains(p);
-        })
-      );
-    });
-
-    return addFilterAsLabelFilter(query, lastPositionsPerExpression, filter);
-  } else {
+  } else if (labelType === LabelType.Indexed) {
     return addFilterToStreamSelector(query, streamSelectorPositions, filter);
+  } else {
+    // labelType is not set, so we need to figure out where to add the label
+    // if we don't have a parser, or have empty stream selectors, we will just add it to the stream selector
+    if (parserPositions.length === 0 || everyStreamSelectorHasMatcher === false) {
+      return addFilterToStreamSelector(query, streamSelectorPositions, filter);
+    } else {
+      // in case we are not adding the label to stream selectors we need to find the last position to add in each expression
+      const subExpressions = findLeaves(getNodePositionsFromQuery(query, [Expr]));
```

Comment:
Not related to this PR, but why do we have this (`const subExpressions = findLeaves(getNodePositionsFromQuery(query, [Expr]));...`) logic in here and not just the same thing as in 
`const positionToAdd = findLastPosition([...streamSelectorPositions, ...labelFilterPositions, ...parserPositions]);`?

Reply:
I'll add a comment that explains this. Basically we need to find the last position in each `Expr`. Because for e.g. `sum(rate({foo="bar"} |logfmt [$__auto])) / sum(rate({foo="baz"}  |logfmt  [$__auto]))` we need to find the last parser position in each sub query, not the overall last position.

Reply:
Added in 3250b7bf1336402ac50ecda25a9a6b868b0027db